### ISSUE TITLE
idomysql fails setting the password

### ISSUE THREAD

**Issue by hboetes** [voxpupuli/puppet-icinga2#295]
Evironment: puppet-4.10.0, all relevant modules come from git, OS: ubuntu-16.04.02

I have this code, or rather a part of the whole code, but this is the relevant part for this bug:

```
  $ic2db     = 'icinga2'                                                                                                                                                                       
  $ic2dbuser = 'icinga2'                                                                                                                                                                       
  $ic2dbpass = 'secret1'                                                                                                                                                                       
   class{ '::icinga2::feature::idomysql':                                                                                                                                                       
    database      => $ic2db,                                                                                                                                                                   
    user          => $ic2dbuser,                                                                                                                                                               
    password      => $ic2dbpass,                                                                                                                                                               
    import_schema => true,                                                                                                                                                                     
    require       => Mysql::Db[$ic2db],                                                                                                                                                        
  }                                                                                                                                                                                            
```                                                                                                                                                                                               

I got this debug output:

```
May  5 15:21:54 icinga02 puppet-agent[21407]: dbconfig-common: writing config to /etc/dbconfig-common/icinga2-ido-mysql.conf
May  5 15:21:54 icinga02 puppet-agent[21407]: creating database backup in /var/cache/dbconfig-common/backups/icinga2-ido-mysql_2.4.1-2ubuntu1.2017-05-05-15.21.53.
May  5 15:21:54 icinga02 puppet-agent[21407]: mysql: [Warning] mysql: Empty value for 'port' specified. Will throw an error in future versions
May  5 15:21:54 icinga02 puppet-agent[21407]: ERROR 1045 (28000): Access denied for user 'icinga2'@'localhost' (using password: YES).
May  5 15:21:54 icinga02 puppet-agent[21407]: unable to connect to mysql server.
May  5 15:21:54 icinga02 puppet-agent[21407]: error encountered backing up the old database:
May  5 15:21:54 icinga02 puppet-agent[21407]: mysql: [Warning] mysql: Empty value for 'port' specified. Will throw an error in future versions ERROR 1045 (28000): Access denied for user 'icin
ga2'@'localhost' (using password: YES)
May  5 15:21:54 icinga02 puppet-agent[21407]: dbconfig-common: icinga2-ido-mysql configure: aborted.
May  5 15:21:54 icinga02 puppet-agent[21407]: dbconfig-common: flushing administrative password
May  5 15:21:54 icinga02 puppet-agent[21407]: dpkg: error processing package icinga2-ido-mysql (--configure):
```
Reading the config file mentioned in the log:
```
# cat /etc/icinga2/features-available/ido-mysql.conf
/**
 * The db_ido_mysql library implements IDO functionality
 * for MySQL.
 */

library "db_ido_mysql"

object IdoMysqlConnection "ido-mysql" {
  user = "icinga2",
  password = "dG6coWH6uyQL",
  host = "localhost",
  database = "icinga2"
}
```

Yes, that's a random password, not the password I set for the dbase.

I suspect this is caused by `spec/classes/idomysql_spec.rb` not having code to set the password, but I don't really understand the code so I can't tell for sure. Could somebody have a look at it?

**Comment by hboetes** [comment]
The plot thickens, after the next puppet run the password is corrected in the beforementioned config-file, but I still get the following error:

```
Error: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install icinga2-ido-mysql' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
icinga2-ido-mysql is already the newest version (2.6.3-1~ppa1~xenial1).
0 upgraded, 0 newly installed, 0 to remove and 25 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Setting up icinga2-ido-mysql (2.6.3-1~ppa1~xenial1) ...
Determining localhost credentials from /etc/mysql/debian.cnf: succeeded.
dbconfig-common: writing config to /etc/dbconfig-common/icinga2-ido-mysql.conf
creating database backup in /var/cache/dbconfig-common/backups/icinga2-ido-mysql_2.4.1-2ubuntu1.2017-05-05-15.39.49.
mysql: [Warning] mysql: Empty value for 'port' specified. Will throw an error in future versions
ERROR 1045 (28000): Access denied for user 'icinga2'@'localhost' (using password: YES).
unable to connect to mysql server.
```

And the database is really configured with the right credentials; `mysql -psecret1 -u icinga2 -h 127.0.0.1 -P3306` succeeds.

**Comment by hboetes** [comment]
The config file `/etc/dbconfig-common/icinga2-ido-mysql.conf` still contains the random password, after replacing it with the right password and running puppet again I finally got a good result.

**Comment by hboetes** [comment]
So please fix this bug and set the right password in right files!

**Comment by bobapple** [comment]
Well.. this is not the default behaviour. Puppet does not use dbconfig when installing packages. Probably you have misconfigured something in the apt module. Please review your code.

**Comment by hboetes** [comment]
I haven’t changed anything in my apt module. It’s completely standard. The whole system is a fresh default install. dbconfig is used by apt on ubuntu for configuring that stuff. Please reopen the bug.

**Comment by hboetes** [comment]
As you can see the the ido-mysql package provided by **your** repository uses dbconfig in the postinstall script.

```
# dpkg -l icinga2-ido-mysql 
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                      Version                   Architecture              Description
+++-=========================================-=========================-=========================-========================================================================================
ii  icinga2-ido-mysql                         2.6.3-1~ppa1~xenial1      amd64                     host and network monitoring system - MySQL support
```
 
```
# cat  /var/lib/dpkg/info/icinga2-ido-mysql.postinst
#!/bin/sh
set -e
. /usr/share/debconf/confmodule
. /usr/share/dbconfig-common/dpkg/postinst
#dbc_debug='1'
dbc_generate_include='template:/etc/icinga2/features-available/ido-mysql.conf'
dbc_generate_include_owner='nagios:nagios'
dbc_generate_include_args='--ucf -o template_infile=/usr/share/icinga2/tmpl/ido-mysql.conf-sample'
dbc_go icinga2-ido-mysql $@
if [ "$1" = configure ]; then
    # get values from debconf db
    db_get icinga2-ido-mysql/enable
    if [ "$RET" = "false" ]; then
        if [ -L /etc/icinga2/features-enabled/ido-mysql.conf ]
        then
            echo "disable ido mysql feature"
            icinga2 feature disable ido-mysql
            echo "reloading icinga2"
            [ -x $(which invoke-rc.d) ] && invoke-rc.d icinga2 reload
        fi
    else
        if [ ! -L /etc/icinga2/features-enabled/ido-mysql.conf ]
        then
            echo "enable ido mysql feature"
            icinga2 feature enable ido-mysql
            echo "reloading icinga2"
            [ -x $(which invoke-rc.d) ] && invoke-rc.d icinga2 reload

        fi
    fi
fi
db_stop
/var/lib/dpkg/info/icinga2-ido-mysql.postinst (END)
```

**Comment by lbetz** [comment]
# puppet apply examples/init_idomysql.pp

works pretty fine on my debian jessie, see below:

Notice: Compiled catalog for debian8.openvpn in environment production in 6.71 seconds
Notice: /Stage[main]/Mysql::Server::Config/File[/etc/mysql]/ensure: created
Notice: /Stage[main]/Mysql::Server::Config/File[/etc/mysql/conf.d]/ensure: created
Notice: /Stage[main]/Mysql::Server::Config/File[mysql-config-file]/ensure: defined content as '{md5}44e7aa974ab98260d7d013a2087f1c77'
Notice: /Stage[main]/Mysql::Server::Install/Package[mysql-server]/ensure: created
Notice: /Stage[main]/Apt/File[/etc/apt/apt.conf.d/15update-stamp]/ensure: defined content as '{md5}4355b3e7824866a503fc221621fc65ba'
Notice: /Stage[main]/Mysql::Client::Install/Package[mysql_client]/ensure: created
Notice: /Stage[main]/Main/Mysql::Db[icinga2]/Mysql_database[icinga2]/ensure: created
Notice: /Stage[main]/Main/Mysql::Db[icinga2]/Mysql_user[icinga2@localhost]/ensure: created
Notice: /Stage[main]/Main/Mysql::Db[icinga2]/Mysql_grant[icinga2@localhost/icinga2.*]/ensure: created
Notice: /Stage[main]/Icinga2::Feature::Idomysql/Package[icinga2-ido-mysql]/ensure: created
Notice: /Stage[main]/Icinga2::Feature::Idomysql/Exec[idomysql-import-schema]/returns: executed successfully
Notice: /Stage[main]/Apt::Backports/Apt::Pin[backports]/File[backports.pref]/ensure: defined content as '{md5}cbe93f6e39e44610f5b39ba0b739285c'
Notice: /Stage[main]/Apt::Backports/Apt::Source[backports]/File[backports.list]/ensure: defined content as '{md5}0209fc2e38ed00d601cabb24cbe85bc4'
Notice: /Stage[main]/Icinga2::Repo/Apt::Source[icinga-stable-release]/Apt::Key[Add key: F51A91A5EE001AA5D77D53C4C6E319C334410682 from Apt::Source icinga-stable-release]/Apt_key[Add key: F51A91A5EE001AA5D77D53C4C6E319C334410682 from Apt::Source icinga-stable-release]/ensure: created
Notice: /Stage[main]/Icinga2::Repo/Apt::Source[icinga-stable-release]/File[icinga-stable-release.list]/ensure: defined content as '{md5}b5de7eef097fa4152f28d09f8cf896b1'
Notice: /Stage[main]/Apt::Update/Exec[apt_update]: Triggered 'refresh' from 2 events
Notice: /Stage[main]/Icinga2::Install/Package[icinga2]/ensure: created
Notice: /Stage[main]/Icinga2::Config/File[/etc/icinga2/constants.conf]/content: content changed '{md5}dee70de6a6961cb6f0815dde93c47775' to '{md5}357d29e0b1deabca26b56c16962445e8'
Notice: /Stage[main]/Icinga2::Config/File[/etc/icinga2/icinga2.conf]/content: content changed '{md5}35da5eaf4658bec1c7a592e34e2f1f20' to '{md5}111f239459c3d62be510ad7c6a9da35c'
Notice: /Stage[main]/Icinga2::Feature::Checker/Icinga2::Object[icinga2::object::CheckerComponent::checker]/Concat[/etc/icinga2/features-available/checker.conf]/File[/etc/icinga2/features-available/checker.conf]/content: content changed '{md5}0b9765380d47ead166be355a33f7f91c' to '{md5}af2f2bbc3c17a3aa199bab5587c2ee99'
Notice: /Stage[main]/Icinga2::Feature::Checker/Icinga2::Object[icinga2::object::CheckerComponent::checker]/Concat[/etc/icinga2/features-available/checker.conf]/File[/etc/icinga2/features-available/checker.conf]/owner: owner changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Checker/Icinga2::Object[icinga2::object::CheckerComponent::checker]/Concat[/etc/icinga2/features-available/checker.conf]/File[/etc/icinga2/features-available/checker.conf]/group: group changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Checker/Icinga2::Object[icinga2::object::CheckerComponent::checker]/Concat[/etc/icinga2/features-available/checker.conf]/File[/etc/icinga2/features-available/checker.conf]/mode: mode changed '0644' to '0640'
Notice: /Stage[main]/Icinga2::Feature::Checker/Icinga2::Feature[checker]/File[/etc/icinga2/features-enabled/checker.conf]/owner: owner changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Checker/Icinga2::Feature[checker]/File[/etc/icinga2/features-enabled/checker.conf]/group: group changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Mainlog/Icinga2::Object[icinga2::object::FileLogger::mainlog]/Concat[/etc/icinga2/features-available/mainlog.conf]/File[/etc/icinga2/features-available/mainlog.conf]/content: content changed '{md5}2660748358a645a59f6b46fa383fa3fa' to '{md5}6938e6b7cdb66a6fc3f0bd3ccc3040de'
Notice: /Stage[main]/Icinga2::Feature::Mainlog/Icinga2::Object[icinga2::object::FileLogger::mainlog]/Concat[/etc/icinga2/features-available/mainlog.conf]/File[/etc/icinga2/features-available/mainlog.conf]/owner: owner changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Mainlog/Icinga2::Object[icinga2::object::FileLogger::mainlog]/Concat[/etc/icinga2/features-available/mainlog.conf]/File[/etc/icinga2/features-available/mainlog.conf]/group: group changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Mainlog/Icinga2::Object[icinga2::object::FileLogger::mainlog]/Concat[/etc/icinga2/features-available/mainlog.conf]/File[/etc/icinga2/features-available/mainlog.conf]/mode: mode changed '0644' to '0640'
Notice: /Stage[main]/Icinga2::Feature::Mainlog/Icinga2::Feature[mainlog]/File[/etc/icinga2/features-enabled/mainlog.conf]/owner: owner changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Mainlog/Icinga2::Feature[mainlog]/File[/etc/icinga2/features-enabled/mainlog.conf]/group: group changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Notification/Icinga2::Object[icinga2::object::NotificationComponent::notification]/Concat[/etc/icinga2/features-available/notification.conf]/File[/etc/icinga2/features-available/notification.conf]/content: content changed '{md5}cb19335158ad227ac8b32baff5d7298b' to '{md5}e005914c31e3e238d5fafc45787d7eaf'
Notice: /Stage[main]/Icinga2::Feature::Notification/Icinga2::Object[icinga2::object::NotificationComponent::notification]/Concat[/etc/icinga2/features-available/notification.conf]/File[/etc/icinga2/features-available/notification.conf]/owner: owner changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Notification/Icinga2::Object[icinga2::object::NotificationComponent::notification]/Concat[/etc/icinga2/features-available/notification.conf]/File[/etc/icinga2/features-available/notification.conf]/group: group changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Notification/Icinga2::Object[icinga2::object::NotificationComponent::notification]/Concat[/etc/icinga2/features-available/notification.conf]/File[/etc/icinga2/features-available/notification.conf]/mode: mode changed '0644' to '0640'
Notice: /Stage[main]/Icinga2::Feature::Notification/Icinga2::Feature[notification]/File[/etc/icinga2/features-enabled/notification.conf]/owner: owner changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Notification/Icinga2::Feature[notification]/File[/etc/icinga2/features-enabled/notification.conf]/group: group changed 'root' to 'nagios'
Notice: /Stage[main]/Icinga2::Feature::Idomysql/Icinga2::Object[icinga2::object::IdoMysqlConnection::ido-mysql]/Concat[/etc/icinga2/features-available/ido-mysql.conf]/File[/etc/icinga2/features-available/ido-mysql.conf]/content: content changed '{md5}5bad48d55764e71b0b56d93ba902ae53' to '{md5}1daf122bdd77e63a1a26dd4744ab08ac'
Notice: /Stage[main]/Icinga2::Feature::Idomysql/Icinga2::Object[icinga2::object::IdoMysqlConnection::ido-mysql]/Concat[/etc/icinga2/features-available/ido-mysql.conf]/File[/etc/icinga2/features-available/ido-mysql.conf]/mode: mode changed '0600' to '0640'
Notice: /Stage[main]/Icinga2::Feature::Idomysql/Icinga2::Feature[ido-mysql]/File[/etc/icinga2/features-enabled/ido-mysql.conf]/ensure: created
Notice: /Stage[main]/Icinga2::Service/Service[icinga2]/ensure: ensure changed 'stopped' to 'running'
Notice: /Stage[main]/Icinga2/Anchor[::icinga2::end]: Triggered 'refresh' from 1 events
Notice: Applied catalog in 153.78 seconds

# mysql -u icinga2 -psupersecret icinga2
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 85
Server version: 5.5.55-0+deb8u1 (Debian)

Copyright (c) 2000, 2017, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> select count(*) from icinga_services;
+----------+
| count(*) |
+----------+
|       12 |
+----------+
1 row in set (0.01 sec)

**Comment by hboetes** [comment]
I just checked with a minimal recipe and that works, do note that `/etc/dbconfig-common/icinga2-ido-mysql.conf` is still generated with a wrong password. But my full recipe fails and I can't see why because it basically does the same. I included it for reference. This recipe includes a workaround which makes it work, but if you comment out the dbconfig workaround I hope you  can reproduce the problem I'm running into. All I changed is the company name and city with sed.
[init.pp.txt](https://github.com/Icinga/puppet-icinga2/files/992351/init.pp.txt)
